rfcs/047-nested-columns.md

```diff
@@ -0,0 +1,189 @@
+2. Properly display nested columns’ metadata (including description, badges, stats) and allow users to edit them
+3. Distinguish how we display nested object without named fields (e.g. Array and Map) vs. a nested object with named fields (e.g. Struct/Message)
+
+
+
+## Motivation
+
+As we look to add more metadata (such as description and stats) for nested columns, the ⌄ icon to expand column details will clash with current use of ↳ icon next to it, making the indentation for hierarchy hard to read and crowding the control area. (See more rationale for overall layout and interaction changes in the [Column details access and side panels RFC](./046-column-details-access-and-side-panels.md))
+
+Also, always showing all nested columns can make the list too long to scroll through.
+
+Finally, there are instances where the nested object we want to display does not have any named fields, such as in Array and Map type nested columns. We will distinguish these from a nested object with named fields (e.g. Struct type nested column), in order to properly represent and highlight the different types of hierarchy. 
+Therefore, this round of enhancements will support use cases like:
+* “I want to put a badge for just this nested column to indicate it is deprecated or reporting safe”
+* “I want to change the description for each level of nested columns”
+We expect these changes will enable our users to navigate and explore nested columns more easily and intuitively, encourage owners to add and maintain higher quality metadata, and help users to clearly understand the hierarchy of all types of nested columns.
+
+
+
+## Guide-level Explanation (aka Product Details)
+
+We have enhanced the way we represent nested columns, including the navigation and a layout change to showcase more metadata per nested columns.
+
+You can expand/collapse all nesting using the text button at top right corner.
+
+Please note that ⌄ icon is now used for expanding each level of nesting, rather than opening the column details within the list. Instead, when you click on the name of any column, you will see all its metadata in the dedicated right side panel. You can always close the panel to see more of the list.
+
+The hierarchy of nested objects without named fields (e.g. Array and Map) are now better specified and distinguished from nested objects with named fields (e.g. Struct).
+
+
+## UI/UX-level Explanation
+
+### Expand/collapse all + each level drill down
+
+![Default all expanded view](../assets/047/default-all-expanded-view.png)
+
+* As default, all nested columns will be expanded, showing the option for user to ‘Collapse all nested’
+* For most complex tables, this default behavior should still load fine, but in the edge case that there are prohibitively many columns to render (threshold will likely be ~1046), we may need to start with an all collapsed view for performance reasons.
```

Comment:
- does it mean that collecting nested column info will be a separate api call to metadata backend?
- how did you come up with this 1046 number?

Reply:
We weren't thinking about changing the API calls.
That number is a bit arbitrary, but it might be the one we are using right now. We'll test it with the worse scenario we have at Lyft and adjust.

Reply:
To expand on the response about the API, the nested column information will be stored as a part of the ColumnMetadata, so it will be available from the existing call to the getTableData API

Reply:
Oops, sorry about the typo on 1046. I meant to say ~1000. We are estimating that even the most complex tables won't have more than 1000 nested columns. This is just to say except for edge cases, we want to show all expanded view as default, but may need special handling for huge tables in case they don't load correctly.